agenda.php

```diff
@@ -9,15 +9,126 @@ class NotImplementedException extends BadMethodCallException {}
+                foreach($vCalendar->VEVENT->ATTENDEE as $attendee) {
+                    if(str_replace("mailto:","", (string)$attendee) === $usermail) {
+                        if(isset($attendee['PARTSTAT']) && (string)$attendee['PARTSTAT'] === "DECLINED") {
+                            $this->log->info("Try to add a user that have already declined invitation (from outside).");
+                            // clean up
+                            $vCalendar->VEVENT->remove($attendee);
+                            // will add again the user
+                            break;
+                        } else {
+                            $this->log->info("Try to add a already registered attendee");
+                            return 0; // not an error
+                        }
+                    }
+                }
+            }
+            
+            $vCalendar->VEVENT->add(
+                'ATTENDEE',
+                'mailto:' . $usermail,
+                [
+                    'CN'   => (is_null($attendee_CN)) ? 'Bénévole' : $attendee_CN,
+                ]
+            );
+            //$vCalendar->VEVENT->add('ATTENDEE', 'mailto:' . $usermail);
+        } else {
+            $already_out = true;
+            
+            if(isset($vCalendar->VEVENT->ATTENDEE)) {
+                foreach($vCalendar->VEVENT->ATTENDEE as $attendee) {
+                    if(str_replace("mailto:","", (string)$attendee) === $usermail) {
+                        $vCalendar->VEVENT->remove($attendee);
+                        $already_out = false;
+                        break;
+                    }
+                }
+            }
+            
+            if($already_out) {
+                $this->log->info("Try to remove an unregistered email");
+                return 0; // not an error
+            }
+        }
+        
+        $new_ETag = $this->caldav_client->updateEvent($vCalendarFilename, $ETag, $vCalendar->serialize());
+        if($new_ETag === false) {
+            $this->log->error("Fails to update the event");
+            return false;
+        } else if(is_null($new_ETag)) {
+            $this->log->info("The server did not answer a new ETag after an event update, need to update the local calendar");
+            if(!$this->updateEvents(array($vCalendarFilename))) {
+                return false;
+            }
+        } else {
+            $this->saveEvent($vCalendarFilename, $new_ETag, $vCalendar->serialize());
+        }
+        return true;
+    }
+    
+    protected function updateEvents(array $vCalendarFilenames) {
+        $events = $this->caldav_client->updateEvents($vCalendarFilenames);
+        
+        if(is_null($events) || $events === false) {
+            $this->log->error("Fail to update events ");
+            return false;
+        }
+
+        foreach($events as $event) {
+            $this->log->info("Adding event $event[vCalendarFilename]");
+            $this->updateEvent($event);
```

Comment:
IIUC this method is only defined on `FSAgenda`. It should probably be declared in this abstract class too

Reply:
Indeed, fixed in b092a7216ddc3cdd5461744c97409cfcc80ab137.
It's weird that PHP does not notice about that ...